task.go

```diff
@@ -200,6 +200,17 @@ func (e *Executor) RunTask(ctx context.Context, call taskfile.Call) error {
 		}
 	}
 
+	// When using the "dir:" attribute it can happen that the directory doesn't exist.
+	// If so, we create it.
+	if t.Dir != "" {
+		if _, err := os.Stat(t.Dir); os.IsNotExist(err) {
+			if err := os.MkdirAll(t.Dir, 0755); err != nil {
```

Comment:
Perhaps move this code to `Executor.Setup()`?

That would prevent race conditions if the same task is ran in parallel.

Reply:
Mhh, I am not very familiar with the internals of task, but here is what I found:

- To create the directory, we need attribute `t.Dir`. At the very beginning of `RunTask()`, we call `CompiledTask()` https://github.com/go-task/task/blob/9c475c36e7f39bca76414c46f783fcb427f905d3/task.go#L178 which can also change the value of `t.Dir`

If we were to move this code to `Setup()`, we would have to do the following:
- Add a loop that goes through all the tasks, so that it can look at each `call` of type `taskfile.Call`, call `t := CompiledTask(call)` and then create the directory based on `t.Dir` ? This would mean also that it would create directories also for tasks that are not required on the command line (directly or indirectly via dependencies).

If my understanding is right, then I need some more guidance on how to do this properly :-)

Reply:
I was also thinking: if there is a race condition in calling `RunTask()`, then this race condition is also there independently from this PR, or am I missing something?

Reply:
Hi @marco-m,

You're right that it'd be more complicated. I totally forgot that we'd need to compile each task.

I think we can then just have a mutex/lock for that. Extracting into another function would be nice, too.

```go
func (t *Task) mkdir() error {
	if t.Dir == "" {
		return nil
	}

	t.mkdirMutex.Lock()
	defer t.mkdirMutex.Unlock()

	// ...
}
```

Reply:
@andreynering I can do that. But what I am missing is this: the mutex would protect for the specific race when creating a directory, but it looks like there is an inherent race when task run in parallel? Wouldn't make more sense to put the mutex outside this new mkdir() function ?